packages/jupyter-ai/jupyter_ai/config_manager.py

```diff
@@ -164,41 +164,48 @@ def _process_existing_config(self, default_config):
                 {k: v for k, v in existing_config.items() if v is not None},
             )
             config = GlobalConfig(**merged_config)
-            validated_config = self._validate_lm_em_id(config)
+            validated_config = self._validate_lm_em_id(
+                config, lm_key="model_provider_id", em_key="embeddings_provider_id"
+            )
+            validated_config = self._validate_lm_em_id(
+                config,
+                lm_key="completions_model_provider_id",
+                em_key="completions_embeddings_provider_id",
+            )
 
             # re-write to the file to validate the config and apply any
             # updates to the config file immediately
             self._write_config(validated_config)
 
-    def _validate_lm_em_id(self, config):
-        lm_id = config.model_provider_id
-        em_id = config.embeddings_provider_id
+    def _validate_lm_em_id(self, config, lm_key, em_key):
+        lm_id = getattr(config, lm_key)
+        em_id = getattr(config, em_key)
 
         # if the currently selected language or embedding model are
         # forbidden, set them to `None` and log a warning.
         if lm_id is not None and not self._validate_model(lm_id, raise_exc=False):
             self.log.warning(
                 f"Language model {lm_id} is forbidden by current allow/blocklists. Setting to None."
             )
-            config.model_provider_id = None
+            setattr(config, lm_key, None)
         if em_id is not None and not self._validate_model(em_id, raise_exc=False):
             self.log.warning(
                 f"Embedding model {em_id} is forbidden by current allow/blocklists. Setting to None."
             )
-            config.embeddings_provider_id = None
+            setattr(config, em_key, None)
 
         # if the currently selected language or embedding model ids are
         # not associated with models, set them to `None` and log a warning.
         if lm_id is not None and not get_lm_provider(lm_id, self._lm_providers)[1]:
             self.log.warning(
                 f"No language model is associated with '{lm_id}'. Setting to None."
             )
-            config.model_provider_id = None
+            setattr(config, lm_key, None)
         if em_id is not None and not get_em_provider(em_id, self._em_providers)[1]:
             self.log.warning(
                 f"No embedding model is associated with '{em_id}'. Setting to None."
             )
-            config.embeddings_provider_id = None
+            setattr(config, em_key, None)
```

Comment:
> Replacing all direct attribute accesses with getattr() and setattr() breaks IDE suggestions in this method, which makes it too easy for developers to introduce bugs into this method in the future.

This is not inherent to use of `setattr`/`getattr`; if you wish to have IDE suggestions, the solution for modern language servers is having a type annotation added in the class body, like:

```python
class ConfigManager:
    model_provider_id: Union[str, None]
    embeddings_provider_id: Union[str, None]
    completions_provider_id: Union[str, None]   # or `Optional[str]`
```

> We should not require calling this method twice to validate the model IDs in the config.

Right, this could be solved by having:

```python
def _validate_lm_em_id(self, config):
   partially_validated_config = self._validate_lm_em_id_pair(
        config,
        lm_key="model_provider_id",
        em_key="embeddings_provider_id"
    )
    validated_config = self._validate_lm_em_id_pair(
        partially_validated_config,
        lm_key="completions_model_provider_id",
        em_key="completions_embeddings_provider_id",
    )
    return validated_config
```

Reply:
Commit https://github.com/jupyterlab/jupyter-ai/pull/711/commits/988213457371050f8f27198af5fb2846b93b705d reverted changes to `_process_existing_config()` and added typing annotations; subsequent commit https://github.com/jupyterlab/jupyter-ai/pull/711/commits/ac7cd846fcd39019da583094c96d825ce8d3922e removed embedding completion models, and finally https://github.com/jupyterlab/jupyter-ai/pull/711/commits/b06481a75d9f9dce9a74e8a83081ffcb96000216 renamed this method to `_validate_model_ids()`.

How does it look for you now?

Reply:
The new implementation looks much cleaner, thank you! I'm fine with using `getattr()` and `setattr()` as long as it's very clear what attributes we are getting/setting.